### ISSUE TITLE
AutoStairs?

### ISSUE THREAD

**Issue by Neubulae** [Andrew6rant/Auto-Slabs#6]
Vertical Stairs can be fun too!

**Comment by Andrew6rant** [comment]
An AutoStairs mod is something that I have been considering, actually.

My problem is, I don't know how to implement it in a way that feels nice to play with.

I could go the easy route and just register a new `SidewaysStairBlock` for every instance of `StairBlock`. I don't want to do that, because I don't like having two blocks for different orientations.

That's actually the reason I made AutoSlabs in the first place - I wasn't happy with how current mods made their vertical slabs separate. I mean, imagine if Minecraft had "top slab" and a "bottom slab" blocks, instead of just one. Not as fun to play with.

On the other side, if I inject blockstates into stair blocks, I can't see any way to implement intuitive placement. Slabs I can do, since they're not based on the player's direction, but Stairs have complex placing (and connecting!) logic.

There are just too many state combinations to nicely put into the block render overlay.

If you have any thoughts on how to alleviate this, by all means, let me know.

**Comment by DadoIrie** [comment]
I don't  mean copy from it, but it might help for inspiration ;) [Additional Placements](https://www.curseforge.com/minecraft/mc-mods/additional-placements)
Is also MIT license this time
Again I'm not suggesting to copy and paste as this not intuitive, but looking on how others resolved a problem can help.

**Comment by Ultimushadow** [comment]
The suggestion from dadorie is a good one, to be honest as a builder I _hate_ that stairs are oriented based on player facing but that implementation seems workable. Maybe you also add a keybind to lock orientation to the last block placed, so it feels nicer to place large amounts of blocks without having to be precise on your orientation or mouse placement? I feel like this could be its own separate suggestion, but when combined with the Additional Placements inspiration I think it could make placing both stairs _and_ slabs better.

**Comment by Andrew6rant** [comment]
A keybind is a nice idea but I feel it isn't the most naturally intuitive. What I'm thinking about is a block placement preview. Here's what I've been playing around with:

![AutoStairs_placement_preview](https://github.com/Andrew6rant/Auto-Slabs/assets/57331134/1a8dc643-1fe3-42e0-8761-5af5949d504a)

![AutoStairs_placement_preview_2](https://github.com/Andrew6rant/Auto-Slabs/assets/57331134/e57465ee-b6e1-4c78-ac10-4589d223cd91)

**Comment by Ultimushadow** [comment]
Interesting, I do think that will really help. What would the vertical stair preview look like then? Two vertical sections in the middle square which would place the corner of the stair on the left/right side of the block you're placing it on? What about the sections when placing a block on the up or down faces of a block?
I threw in the keybind as a fix for the tedium of placing multiple stair blocks with the same rotation and how the player in vanilla has to be precise in their position. While this gets around the position issue since you can place on diagonals, it's traded for mouse precision. That may not seem like a big deal but as someone who struggles with muscle/wrist issues, it can become a massive pain, literally, to place a long line of blocks precisely. The keybind would merely be for accessibility and not the default in my suggestion, and it would only temporarily save the orientation (N/E/S/W and top/bottom), not the shape. While I know that giving players options leads to wasted effort in many mods, I have a feeling all builders would especially enjoy this feature, even if they don't have physical troubles, I don't know anyone who _enjoys_ having to break all the misplaced blocks while building a roof or something similar 😅

**Comment by Andrew6rant** [comment]
Yeah, there are too many stair states to nicely put in the block render overlay.

I have been thinking about a "lock direction" keybind, as well as "cycle left" and "cycle right" keybinds, for easier placement.

The challenge with that is that I need to do networking. Slab placement is done using a server-sided raycast. I need to let the server know what client bind is pressed, otherwise it would just use the player's position to calculate the placed state.

**Comment by Ultimushadow** [comment]
The Additional Placements set up looked pretty clean but to be honest, I think what you're looking into now is just flat out better. Keybind cycling makes a lot of sense, would it be possible to implement a way to use the scroll wheel for that too? Maybe when rotation is locked, holding shift enables the scroll wheel to quickly change it?

Yeah, when you mentioned the way the mod implements things server side on the Fluidlogged issue report, it put some pieces together lol
Hopefully the communication doesn't give you too much trouble - I did have a question though. Why does the client lack the block placing animation with this mod? Shouldn't that be all client side?

**Comment by Andrew6rant** [comment]
Yeah I can look into doing that. I can even try to make a little render above the crosshair with the placed slab wireframe, just so it is easier to tell which state you've locked to.

> Why does the client lack the block placing animation with this mod? Shouldn't that be all client side?

I set the placement state on the client to null, and let the server handle all placement code. I honestly didn't notice that there was no animation, haha. The reason is that if the player moves quickly while placing a slab or a row of slabs, the server and client raycasts (which are triggered at slightly different times) can differ, causing visual desync.

The only way to fix this (I thought) would be to sync the client placement code with the server. I actually found a workaround, though.

```diff
- if (!(ctx.getWorld() instanceof ServerWorld)) return null;
+ if (!(ctx.getWorld() instanceof ServerWorld)) return Blocks.LIGHT.getDefaultState().with(LEVEL_15, 0);
```

Instead of returning null, I return an entirely invisible block with no collision. This sidesteps my need for networking (since the client thinks it places a block and the server handles the placement), and actually fixes #14! (since I am no longer returning null)

There is a potential for breakage if a mod assumes that the result of a SlabBlock's `getPlacementState` is an instanceof SlabBlock on the client (because LightBlocks are not slabs), but I doubt that would happen because any mod that needs to do that should hopefully check on the server and not both sides or the client.

The only mod that I can think of that is affected by my change is actually another one of my own, https://github.com/Andrew6rant/Custom-Placement-Preview. I used to rely on the client placement state to render my slab wireframe, but I have an API for compatibility and am going to use that instead when installed alongside AutoSlabs.

**Comment by Andrew6rant** [comment]
Implemented a keybind system that allows you to cycle through slab placement states (use shift + keybind to loop in reverse)

![autoslabs_placement_lock_gui](https://github.com/Andrew6rant/Auto-Slabs/assets/57331134/50795a62-36a3-4f1b-915f-e77b610a8d0b)

Textures are WIP (and I haven't done the lang yet), but the render works

**Comment by Anoaxx** [comment]
Take a connection logic from stairs and implement it to vertical slab? like sticking to each other from outside and inside, like if they are placed next to each other but facing other directions they can connect into vertical stair or vertical corner? The problem is u cant have just sideways stairs, but if u would somehow make them stay in that position if u break block next to it or just make it possible with debugstick. It could be added for stairs too.

**Comment by thwonn** [comment]
yes vertical corners would be a nightmare
